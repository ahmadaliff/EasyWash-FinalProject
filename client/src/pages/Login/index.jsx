import PropTypes from 'prop-types';
import toast from 'react-hot-toast';
import { useForm } from 'react-hook-form';
import { useEffect, useState } from 'react';
import { connect, useDispatch } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { Link, useNavigate } from 'react-router-dom';
import { FormattedMessage, injectIntl } from 'react-intl';

import { Button, Card, CardContent } from '@mui/material';

import InputRHF from '@components/InputRHF';

import { selectLogin } from '@containers/Client/selectors';
import { actionHandleLogin, actionHandleLoginGoogle } from '@containers/Client/actions';

import classes from '@pages/Login/style.module.scss';
import { Google } from '@mui/icons-material';

const Login = ({ isLogin, intl: { formatMessage } }) => {
  const navigate = useNavigate();
  const dispatch = useDispatch();
  const [showPass, setShowPass] = useState(false);
  const [isAfterLogin, setIsAfterLogin] = useState(false);
  const {
    handleSubmit,
    register,
    formState: { errors },
    watch,
  } = useForm();
  const email = watch('email');
  const password = watch('password');

  useEffect(() => {
    if (isLogin && !isAfterLogin) {
      toast.error(formatMessage({ id: 'app_user_already_login' }));
      navigate('/');
    }
  }, [formatMessage, isAfterLogin, isLogin, navigate]);

  const onSubmit = (data) => {
    setIsAfterLogin(true);
    dispatch(
      actionHandleLogin(data, () => {
        setTimeout(() => {
          navigate('/');
        }, 1000);
      })
    );
  };

  return (
    <main className={classes.main}>
      <img src="login.svg" className={classes.loginIlustration} alt="https://storyset.com" />
      <div className={classes.content}>
        <div onClick={() => navigate('/')} className={classes.logoHeader}>
          <img src="longLogo.svg" alt="logo" className={classes.logo} />
        </div>
        <Card className={classes.loginCard}>
          <CardContent>
            <div className={classes.headerLogin}>
              <h3>
                <FormattedMessage id="app_header_login" />
              </h3>
              <Link to="/register">
                <FormattedMessage id="app_header_register" />
              </Link>
            </div>
            <form onSubmit={handleSubmit(onSubmit)}>
              <InputRHF
                input={{
                  name: 'email',
                  required: formatMessage({ id: 'app_user_email_require_message' }),
                  type: 'text',
                  label: formatMessage({ id: 'app_user_email' }),
                  pattern: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+.[A-Z]{2,}$/i,
                  messagePatern: formatMessage({ id: 'app_user_email_pattern_message' }),
                }}
                register={register}
                errors={errors}
              />
              <InputRHF
                input={{
                  name: 'password',
                  required: formatMessage({ id: 'app_user_password_require_message' }),
                  type: showPass ? 'text' : 'password',
                  label: formatMessage({ id: 'app_user_password' }),
                  minLength: 8,
                  messageMin: formatMessage({ id: 'app_user_password_min_length' }),
                }}
                register={register}
                errors={errors}
              >
                <label htmlFor="show" className={classes.showPassword}>
                  <input type="checkbox" name="show" id="show" onChange={(e) => setShowPass(e.target.checked)} />
                  <FormattedMessage id="app_user_password_show" />
                </label>
              </InputRHF>
              <p className={classes.nav}>
                <FormattedMessage id="app_forgot_password" /> ?
                <Link to="/forgotPassword">
                  <FormattedMessage id="app_forgot_password" />
                </Link>
              </p>
              <button type="submit" className={classes.buttonSubmit} disabled={!email || !password}>
                <FormattedMessage id="app_header_login" />
              </button>
            </form>
            <p className={classes.or}>
              <FormattedMessage id="app_or_login" />
            </p>
            <Button
              className={classes.buttonGoogle}
              onClick={() => dispatch(actionHandleLoginGoogle())}
              variant="outlined"
              startIcon={<Google />}
              fullWidth
            >
              Google
            </Button>
          </CardContent>
        </Card>
        <a href="https://storyset.com/web" className={classes.storySet}>
          Web illustrations by Storyset
        </a>
      </div>
    </main>
  );
};

Login.propTypes = {
  intl: PropTypes.object,
  isLogin: PropTypes.bool,
};

const mapStateToProps = createStructuredSelector({
  isLogin: selectLogin,
});

export default injectIntl(connect(mapStateToProps)(Login));
