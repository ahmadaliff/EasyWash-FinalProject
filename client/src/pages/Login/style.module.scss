@import '@styles/base';
.main {
  @include directionY;
  position: relative;
  min-height: 100vh;
  align-items: center;
  @include apply-at-sm-size {
    justify-content: center;
  }
  .loginIlustration {
    position: absolute;
    height: 0;
    max-width: 100vw;
    @include apply-at-sm-size {
      height: 50vh;
    }
    @include apply-at-md-size {
      height: 90vh;
    }
  }
  .storySet {
    display: none;
    @include apply-at-sm-size {
      display: block;
    }
    margin: 0.5rem;
    font-size: x-small;
    color: var(--color-primary);
  }
  .content {
    @include directionY;
    align-items: center;
    z-index: 2;
    width: 100%;
    .logoHeader {
      position: absolute;
      cursor: pointer;
      width: auto;
      top: 0;
      padding: 1rem;
      height: 2.5rem;
      .logo {
        height: 2.5rem;
      }
    }
    .loginCard {
      padding-top: 5rem;
      width: 100%;
      min-height: 100vh;
      box-shadow: 0 0 0.25rem gray;
      @include apply-at-sm-size {
        margin-top: 1.5rem;
        padding-top: 0;
        min-height: auto;
        height: 100%;
        max-width: 17rem;
      }
      .headerLogin {
        @include directionX;
        align-items: end;
        justify-content: space-between;
        padding: 0 0 1.5rem 0;
        h3 {
          margin: 0;
        }
        a {
          font-size: small;
          color: var(--color-primary);
          text-decoration: none;
        }
      }
      .showPassword,
      .nav {
        @include directionX;
        align-items: center;
        padding-top: 0.25rem;
        font-size: small;
        color: var(--color-primary);
      }
      .nav {
        color: var(--color-text);
        a {
          color: var(--color-primary);
        }
      }
      .buttonSubmit {
        width: 100%;
        background-color: var(--color-primary);
        color: var(--color-bg);
        border-radius: 2.5rem;
        padding: 0.5rem 2.5rem 0.5rem 2.5rem;
        margin: auto;
        margin-top: 0.75rem;
        box-shadow: none;
        border: none;
        cursor: pointer;
        &:hover {
          transform: scale(1.0125, 1.0125);
        }
        &:disabled {
          background-color: rgba(gray, 0.2);
          color: rgba($color: gray, $alpha: 0.5);
        }
        &:disabled:hover {
          transform: none;
        }
      }
      .or {
        font-size: small;
        text-align: center;
      }
      .buttonGoogle {
        text-transform: capitalize;
        color: var(--color-primary);
        border-color: var(--color-primary);
        border-radius: 20px;
      }
    }
  }
}
