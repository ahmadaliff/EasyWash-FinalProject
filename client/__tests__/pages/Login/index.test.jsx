import Login from '@pages/Login';
import { act, fireEvent, render, screen } from '@utils/testHelper';

let wrapper;
const mockNavigate = jest.fn();
const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDipatch: () => mockDispatch,
}));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
  Link: () => <b />,
}));

beforeEach(() => {
  act(() => {
    wrapper = render(<Login />);
  });
});

describe('Login Page', () => {
  test('render', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('login-wrap')).toBeInTheDocument();
  });
  test('shuldbe navigate to home', async () => {
    const { getByTestId } = wrapper;
    const button = getByTestId('button-logo');
    act(() => {
      fireEvent.click(button);
    });
    expect(mockNavigate).toHaveBeenCalledWith('/');
  });
  test('shouldbe running on submit function', async () => {
    const { getByTestId } = wrapper;
    const inputEmail = screen.getByLabelText('Email');
    const inputPassword = screen.getByLabelText('Kata Sandi');
    fireEvent.change(inputEmail, { target: { value: 'email@email.com' } });
    fireEvent.change(inputPassword, { target: { value: 'password123' } });
    const button = getByTestId('button-submit');
    await act(async () => {
      fireEvent.click(button);
      fireEvent.click(getByTestId('button-google'));
    });
  });
  test('shouldbe change type input password', () => {
    const { getByTestId } = wrapper;
    const button = getByTestId('button-show');
    act(() => {
      fireEvent.click(button);
    });
  });
  test('Matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
