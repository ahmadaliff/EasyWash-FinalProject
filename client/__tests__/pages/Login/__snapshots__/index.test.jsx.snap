// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page Matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="main"
        data-testid="login-wrap"
      >
        <img
          alt="https://storyset.com"
          class="loginIlustration"
          src="login.svg"
        />
        <div
          class="content"
        >
          <div
            class="logoHeader"
            data-testid="button-logo"
          >
            <img
              alt="logo"
              class="logo"
              src="longLogo.svg"
            />
          </div>
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root loginCard css-bhp9pd-MuiPaper-root-MuiCard-root"
          >
            <div
              class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
            >
              <div
                class="headerLogin"
              >
                <h3>
                  Masuk
                </h3>
                <b />
              </div>
              <form>
                <div
                  class="MuiFormControl-root inputContainer css-1nrlq1o-MuiFormControl-root"
                  data-testid="inputRHF"
                >
                  <label
                    class="inputLabel"
                    for="email"
                  >
                    Email
                    <input
                      class="input undefined"
                      id="email"
                      name="email"
                      type="text"
                      value=""
                    />
                  </label>
                </div>
                <div
                  class="MuiFormControl-root inputContainer css-1nrlq1o-MuiFormControl-root"
                  data-testid="inputRHF"
                >
                  <label
                    class="inputLabel"
                    for="password"
                  >
                    Kata Sandi
                    <input
                      class="input undefined"
                      id="password"
                      name="password"
                      type="password"
                      value=""
                    />
                  </label>
                  <label
                    class="showPassword"
                    for="show"
                  >
                    <input
                      data-testid="button-show"
                      id="show"
                      name="show"
                      type="checkbox"
                    />
                    Tampilkan Kata Sandi
                  </label>
                </div>
                <p
                  class="nav"
                >
                  Lupa Password
                   ?
                  <b />
                </p>
                <button
                  class="buttonSubmit"
                  data-testid="button-submit"
                  disabled=""
                  type="submit"
                >
                  Masuk
                </button>
              </form>
              <p
                class="or"
              >
                atau masuk dengan
              </p>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-fullWidth buttonGoogle css-xsqe0f-MuiButtonBase-root-MuiButton-root"
                data-testid="button-google"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="GoogleIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12.545,10.239v3.821h5.445c-0.712,2.315-2.647,3.972-5.445,3.972c-3.332,0-6.033-2.701-6.033-6.032s2.701-6.032,6.033-6.032c1.498,0,2.866,0.549,3.921,1.453l2.814-2.814C17.503,2.988,15.139,2,12.545,2C7.021,2,2.543,6.477,2.543,12s4.478,10,10.002,10c8.396,0,10.249-7.85,9.426-11.748L12.545,10.239z"
                    />
                  </svg>
                </span>
                Google
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
          <a
            class="storySet"
            href="https://storyset.com/web"
          >
            Web illustrations by Storyset
          </a>
        </div>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="main"
      data-testid="login-wrap"
    >
      <img
        alt="https://storyset.com"
        class="loginIlustration"
        src="login.svg"
      />
      <div
        class="content"
      >
        <div
          class="logoHeader"
          data-testid="button-logo"
        >
          <img
            alt="logo"
            class="logo"
            src="longLogo.svg"
          />
        </div>
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root loginCard css-bhp9pd-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              class="headerLogin"
            >
              <h3>
                Masuk
              </h3>
              <b />
            </div>
            <form>
              <div
                class="MuiFormControl-root inputContainer css-1nrlq1o-MuiFormControl-root"
                data-testid="inputRHF"
              >
                <label
                  class="inputLabel"
                  for="email"
                >
                  Email
                  <input
                    class="input undefined"
                    id="email"
                    name="email"
                    type="text"
                    value=""
                  />
                </label>
              </div>
              <div
                class="MuiFormControl-root inputContainer css-1nrlq1o-MuiFormControl-root"
                data-testid="inputRHF"
              >
                <label
                  class="inputLabel"
                  for="password"
                >
                  Kata Sandi
                  <input
                    class="input undefined"
                    id="password"
                    name="password"
                    type="password"
                    value=""
                  />
                </label>
                <label
                  class="showPassword"
                  for="show"
                >
                  <input
                    data-testid="button-show"
                    id="show"
                    name="show"
                    type="checkbox"
                  />
                  Tampilkan Kata Sandi
                </label>
              </div>
              <p
                class="nav"
              >
                Lupa Password
                 ?
                <b />
              </p>
              <button
                class="buttonSubmit"
                data-testid="button-submit"
                disabled=""
                type="submit"
              >
                Masuk
              </button>
            </form>
            <p
              class="or"
            >
              atau masuk dengan
            </p>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-fullWidth buttonGoogle css-xsqe0f-MuiButtonBase-root-MuiButton-root"
              data-testid="button-google"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="GoogleIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12.545,10.239v3.821h5.445c-0.712,2.315-2.647,3.972-5.445,3.972c-3.332,0-6.033-2.701-6.033-6.032s2.701-6.032,6.033-6.032c1.498,0,2.866,0.549,3.921,1.453l2.814-2.814C17.503,2.988,15.139,2,12.545,2C7.021,2,2.543,6.477,2.543,12s4.478,10,10.002,10c8.396,0,10.249-7.85,9.426-11.748L12.545,10.239z"
                  />
                </svg>
              </span>
              Google
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <a
          class="storySet"
          href="https://storyset.com/web"
        >
          Web illustrations by Storyset
        </a>
      </div>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
